File: Cpp-C/Ema/Src/Access/Impl/ServiceEndpointDiscoveryInfo.cpp
/*|-----------------------------------------------------------------------------
 *|            This source code is provided under the Apache 2.0 license      --
 *|  and is provided AS IS with no warranty or guarantee of fit for purpose.  --
 *|                See the project's LICENSE.md for details.                  --
 *|           Copyright (C) 2019 Refinitiv. All rights reserved.            --
 *|-----------------------------------------------------------------------------
 */

#include "ServiceEndpointDiscoveryInfo.h"
#include "ExceptionTranslator.h"
#include "Utilities.h"

#include <new>

using namespace rtsdk::ema::access;

ServiceEndpointDiscoveryInfo::ServiceEndpointDiscoveryInfo() :
_toString(0,256)
{
	_pDataFormatList = 0;
	_pLocationList = 0;
	try
	{
		_pDataFormatList = new EmaVector<EmaString>(2);
		_pLocationList = new EmaVector<EmaString>(2);
	}
	catch (std::bad_alloc)
	{
		// it's safe to delete nullptr
		delete _pDataFormatList;
		_pDataFormatList = 0;
		delete _pLocationList;
		_pLocationList = 0;

		throwMeeException("Failed to allocate memory for EmaVector<EmaString> in ServiceEndpointDiscoveryInfo().");
	}
}

ServiceEndpointDiscoveryInfo::~ServiceEndpointDiscoveryInfo()
{
	if (_pDataFormatList)
	{
		delete _pDataFormatList;
		_pDataFormatList = 0;
	}

	if (_pLocationList)
	{
		delete _pLocationList;
		_pLocationList = 0;
	}
}

const EmaVector<EmaString>& ServiceEndpointDiscoveryInfo::getDataFormatList() const
{
	return *_pDataFormatList;
}

const EmaString&	ServiceEndpointDiscoveryInfo::getEndPoint() const
{
	return _endPoint;
}

const EmaVector<EmaString>& ServiceEndpointDiscoveryInfo::getLocationList() const
{
	return *_pLocationList;
}

const EmaString&	ServiceEndpointDiscoveryInfo::getPort() const
{
	return _port;
}

const EmaString&	ServiceEndpointDiscoveryInfo::getProvider() const
{
	return _provider;
}

const EmaString&	ServiceEndpointDiscoveryInfo::getTransport() const
{
	return _transport;
}

ServiceEndpointDiscoveryInfo& ServiceEndpointDiscoveryInfo::operator=(const ServiceEndpointDiscoveryInfo& other)
{
	UInt32 index;
	_pDataFormatList->clear();
	for (index = 0; index < other.getDataFormatList().size(); index++)
	{
		_pDataFormatList->push_back(other.getDataFormatList()[index]);
	}
	
	_endPoint = other.getEndPoint();

	_pLocationList->clear();
	for (index = 0; index < other.getLocationList().size(); index++)
	{
		_pLocationList->push_back(other.getLocationList()[index]);
	}

	_port = other.getPort();
	_provider = other.getProvider();
	_transport = other.getTransport();

	return *this;
}

const EmaString& ServiceEndpointDiscoveryInfo::toString() const
{
	return toString(0);
}

const EmaString& ServiceEndpointDiscoveryInfo::toString(UInt64 indent) const
{
	UInt32 index;
	addIndent(_toString.clear().append("Service : "), indent, true);
	addIndent(_toString.append("Provider : ").append(_provider), indent, true);
	addIndent(_toString.append("Transport : ").append(_transport), indent, true);
	addIndent(_toString.append("Endpoint : ").append(_endPoint), indent, true);
	addIndent(_toString.append("Port : ").append(_port), indent, true);

	_toString.append("Data Format : ");
	for (index = 0; index < _pDataFormatList->size(); index++)
		_toString.append((*_pDataFormatList)[index]).append("  ");

	addIndent(_toString, indent, true);
	_toString.append("Location : ");
	for (index = 0; index < _pLocationList->size(); index++)
		_toString.append((*_pLocationList)[index]).append(" ");

	return _toString;
}

ServiceEndpointDiscoveryInfo::operator const char* () const
{
	return _toString.c_str();
}

File: Java/Ema/Core/src/main/java/com/thomsonreuters/ema/access/ServiceEndpointDiscoveryInfo.java
///*|-----------------------------------------------------------------------------
// *|            This source code is provided under the Apache 2.0 license      --
// *|  and is provided AS IS with no warranty or guarantee of fit for purpose.  --
// *|                See the project's LICENSE.md for details.                  --
// *|           Copyright (C) 2019 Refinitiv. All rights reserved.            --
///*|-----------------------------------------------------------------------------

package com.thomsonreuters.ema.access;

import java.util.List;

/**
 * ServiceEndpointDiscoveryInfo represents an service endpoint information from EDP-RT service discovery.
 *
 * <p>This class is retrieved from ServiceEndpointDiscoveryResp</p>
 * 
 * @see ServiceEndpointDiscoveryResp
 */

public interface ServiceEndpointDiscoveryInfo
{
	/**
	 * Gets a list of data format supported by this endpoint.
	 * 
	 * @return a list of data format
	 */
	public List<String> dataFormatList();
	
	/**
	 * Gets an endpoint or domain name for establishing a connection
	 * 
	 * @return an endpoint
	 */
	public String endpoint();
	
	/**
	 * Gets a list of locations where the infrastructure is deployed in Refinitiv Real-Time Optimized 
	 * 
	 * @return a list of location
	 */
	public List<String> locationList();
	
	/**
	 * Gets a port for establishing a connection
	 * 
	 * @return a port
	 */
	public String port();
	
	/**
	 * Gets a public Refinitiv Real-Time Optimized provider
	 * 
	 * @return a Refinitiv Real-Time Optimized provider
	 */
	public String provider();
	
	
	/**
	 * Gets a transport type
	 * 
	 * @return a transport type
	 */
	public String transport();
	
	/**
     * Gets string representation of this object
     * 
     * @return String representation of ServiceEndpointDiscoveryInfo.
     */
    public String toString();
}
